### PR TITLE
Library.kodi.audioengine

### PR DESCRIPTION
**Pull request by AchimTuran**
@fritsch @FernetMenta 
This PR adds a wrapper library for the binary add-ons, which allows them to create audio streams.

### CODE REVIEW

File: lib/addons/library.kodi.audioengine/libKODI_audioengine.cpp [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+  }
+    
+  return ((CB_AudioEngineLib*)cb)->GetCurrentSinkFormat(((AddonCB*)hdl)->addonData, SinkFormat);
+}
+
+CAddonAEStream::CAddonAEStream(void *Addon, void *Callbacks, AEStreamHandle *StreamHandle)
+{
+  if(!Addon || !Callbacks || !StreamHandle)
+  {
+    // ToDo: throw error!
```
Comment (FernetMenta) [inline lib/addons/library.kodi.audioengine/libKODI_audioengine.cpp]: ?

File: xbmc/addons/AddonCallbacks.cpp [hunk 2]
```diff
@@ -33,26 +34,29 @@ namespace ADDON
...
-  m_callbacks->AddOnLib_UnRegisterMe = CAddonCallbacks::AddOnLib_UnRegisterMe;
-  m_callbacks->ADSPLib_RegisterMe    = CAddonCallbacks::ADSPLib_RegisterMe;
-  m_callbacks->ADSPLib_UnRegisterMe  = CAddonCallbacks::ADSPLib_UnRegisterMe;
-  m_callbacks->CODECLib_RegisterMe   = CAddonCallbacks::CODECLib_RegisterMe;
-  m_callbacks->CODECLib_UnRegisterMe = CAddonCallbacks::CODECLib_UnRegisterMe;
-  m_callbacks->GUILib_RegisterMe     = CAddonCallbacks::GUILib_RegisterMe;
-  m_callbacks->GUILib_UnRegisterMe   = CAddonCallbacks::GUILib_UnRegisterMe;
-  m_callbacks->PVRLib_RegisterMe     = CAddonCallbacks::PVRLib_RegisterMe;
-  m_callbacks->PVRLib_UnRegisterMe   = CAddonCallbacks::PVRLib_UnRegisterMe;
+  m_addon             = addon;
```
Comment (FernetMenta) [inline xbmc/addons/AddonCallbacks.cpp]: I am not a friend of vertical alignment of operators. it just ruins the diff with zero advantage on the other side.

File: lib/addons/library.kodi.audioengine/libKODI_audioengine.cpp (hunk 1 above)
Comment (AchimTuran) [inline lib/addons/library.kodi.audioengine/libKODI_audioengine.cpp]: This is still a to do, but at the moment I don't know if it is necessary. If not I can remove it.

I wanted to throw an exception at this line.

File: xbmc/addons/AddonCallbacks.cpp (hunk 2 above)
Comment (AchimTuran) [inline xbmc/addons/AddonCallbacks.cpp]: No problem. I can remove this.

File: lib/addons/library.kodi.audioengine/libKODI_audioengine.cpp (hunk 1 above)
Comment (FernetMenta) [inline lib/addons/library.kodi.audioengine/libKODI_audioengine.cpp]: you can check those before calling the contructor

Comment (AchimTuran) [inline lib/addons/library.kodi.audioengine/libKODI_audioengine.cpp]: You're right. I will remove this in the final version.

File: xbmc/addons/AddonCallbacks.cpp (hunk 2 above)
Comment (garbear) [inline xbmc/addons/AddonCallbacks.cpp]: or at least make it a separate commit. and of course omit it from the PR, because there's nothing sexier than a pr with 1740 lines added, 2 removed ;)

**Comment by AchimTuran** [comment]
@FernetMenta @fritsch 
This PR is also updated

File: addons/library.kodi.audioengine/libKODI_audioengine.h [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+  bool RegisterMe(void* handle)
+  {
+    m_Handle = handle;
+
+    std::string libBasePath;
+    libBasePath  = ((cb_array*)m_Handle)->libPath;
+    libBasePath += AUDIOENGINE_HELPER_DLL;
+
+#if defined(ANDROID)
+      struct stat st;
```
Comment (FernetMenta) [inline addons/library.kodi.audioengine/libKODI_audioengine.h]: check indent

**Comment by FernetMenta** [comment]
looks ok to me.

**Comment by AchimTuran** [comment]
TODO: update the GetCurrentSinkFormat(...), which is modified in https://github.com/xbmc/xbmc/pull/7609.

**Comment by AchimTuran** [comment]
jenkins build this please

**Comment by AchimTuran** [comment]
jenkins build this please

**Comment by AchimTuran** [comment]
@Memphiz or @FernetMenta 
Can you help to fix OSX and iOS?

**Comment by AchimTuran** [comment]
jenkins build this please

**Comment by Memphiz** [comment]
If you know that osx/ios is broken and you want to fix for example linux only - it might be better to the ressources to only kick a linux build then.

**Comment by FernetMenta** [comment]
@AchimTuran I can have a look this week end.

**Comment by garbear** [comment]
@AchimTuran I didn't get a chance to go in depth, but I have a few fixes you can apply

See https://github.com/garbear/xbmc/compare/38b9a74%5E...1d58b84

The commits were originally fixes for the commit "Basic library.kodi.interfaces implementation" (https://github.com/garbear/xbmc/commit/c948eb5) but it looks like you've split that one up

The first one, "Alphabetize addon callbacks and revert whitespace changes", was needed so I could rebase your PR on my retroplayer branch. After that, it rebased cleanly.

It also looks like you got the linux fix

Hope the rest are helpful. I'll return to this when I can fit it into my game library work

**Comment by AchimTuran** [comment]
> @AchimTuran I can have a look this week end.

Thanks @FernetMenta, that would be nice 

> If you know that osx/ios is broken and you want to fix for example linux only - it might be better to the ressources to only kick a linux build then.

You're right @Memphiz. I'm still learning jenkins, but now I know how to build for a specific platform.

> @AchimTuran I didn't get a chance to go in depth, but I have a few fixes you can apply
See garbear@38b9a74^...1d58b84

> The commits were originally fixes for the commit "Basic library.kodi.interfaces implementation" (garbear@c948eb5) but it looks like you've split that one up

> The first one, "Alphabetize addon callbacks and revert whitespace changes", was needed so I could rebase your PR on my retroplayer branch. After that, it rebased cleanly.

> It also looks like you got the linux fix

> Hope the rest are helpful. I'll return to this when I can fit it into my game library work

Absolutly no problem. The library.kodi.interfaces PR was one of the last things I was working on during GSoC. Currently this PR needs the most work to get it into a good shape. Thanks for your work, I will include them during the next week.
Again @garbear  feel free to ask questions about my work during GSoC.

**Comment by FernetMenta** [comment]
@AchimTuran xcode updated: https://github.com/FernetMenta/xbmc/commit/3ae23b5fd135674098bbdbc19f7bcd5a7976cd60

**Comment by AchimTuran** [comment]
@FernetMenta thx for your help.

jenkins build this please

**Comment by MartijnKaijser** [comment]
Won't work. Needs rebase

**Comment by AchimTuran** [comment]
@MartijnKaijser
> Won't work. Needs rebase

You're right. Sorry I forgot that step.

**Comment by garbear** [comment]
@AchimTuran I'm creating a fixup commit that you can squash into the rebase. Rebasing RetroPlayer on this patch creates a bunch of merge conflicts that can be prevented by alphabetizing some stuff. I'm ok doing work now to prevent headaches later :)

**Comment by AchimTuran** [comment]
@garbear 
> @AchimTuran I'm creating a fixup commit that you can squash into the rebase. Rebasing RetroPlayer on this patch creates a bunch of merge conflicts that can be prevented by alphabetizing some stuff. I'm ok doing work now to prevent headaches later :)

That would be nice and is a good solution.
Anyway I must squash the project files (VS, XCode and Make).

**Comment by garbear** [comment]
here you go: https://github.com/garbear/xbmc/compare/69204c4...4093292

**Comment by AchimTuran** [comment]
@garbear 
That was very fast. It's now integrate. Hopefully you like it.

**Comment by garbear** [comment]
missed a few in .gitignore, but they're not so headache-inducing

**Comment by AchimTuran** [comment]
@garbear 
> missed a few in .gitignore, but they're not so headache-inducing

It's up to you, if you want we can integrate that in this PR.

again jenkins build this please

**Comment by garbear** [comment]
nah go for it

**Comment by MartijnKaijser** [comment]
Might wanna squash that temp linux fix commit

**Comment by AchimTuran** [comment]
> Might wanna squash that temp linux fix commit

Sure, I will also squash the XCode commit.

**Comment by AchimTuran** [comment]
From my side this PR is ready to merge.

The last thing is that yesterday OSX32 failed. Can anyone help?

**Comment by Memphiz** [comment]
i don't see a build log on jenkins / osx32 for this pr ...

**Comment by AchimTuran** [comment]
> i don't see a build log on jenkins / osx32 for this pr ...

It's here: http://jenkins.kodi.tv/job/BuildMulti-PR/3641/

**Comment by AchimTuran** [comment]
jenkins build this please

**Comment by AchimTuran** [comment]
@Memphiz or @FernetMenta or @MartijnKaijser 

Do you know why jenkins OSX64 failed. Seems to be that it crashed before build start.

**Comment by MartijnKaijser** [comment]
not important error. lets try again
jenkins build this please

**Comment by Memphiz** [comment]
build log is not available anymore - but the buildtime was only 36 seconds - this is a github connection issue for sure ... no worries.

**Comment by AchimTuran** [comment]
Now it builds all platforms.

Can we merge this PR?

**Comment by FernetMenta** [comment]
your button

**Comment by AchimTuran** [comment]
@FernetMenta 
thanks for the reminder. Here is the corresponding PR https://github.com/xbmc/xbmc/pull/8117
